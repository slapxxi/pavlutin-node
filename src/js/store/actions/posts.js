import types from './types';

function addPost(post) {
  return {
    type: types.ADD_POST,
    post,
  };
}

function requestPosts() {
  return {
    type: types.REQUEST_POSTS,
  };
}

function receivePosts(json) {
  return {
    type: types.RECEIVE_POSTS,
    posts: json.posts,
    receivedAt: Date.now(),
  };
}

function fetchPosts() {
  return function fetchPostsThunk(dispatch) {
    dispatch(requestPosts());
    return fetch('/api/v1/posts')
      .then(r => r.json())
      .then(json => dispatch(receivePosts(json)));
  };
}

export { addPost, requestPosts, receivePosts, fetchPosts };
