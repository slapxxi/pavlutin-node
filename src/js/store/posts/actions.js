// @flow
import PostsAPI from '../../services/PostsAPI';
import type { Action, Post, Dispatch } from '../flow-types';

function addPost(post: Post): Action {
  return {
    type: 'ADD_POST',
    payload: post,
  };
}

function requestPosts(): Action {
  return {
    type: 'REQUEST_POSTS',
  };
}

function requestPostsError(e: Error): Action {
  return {
    type: 'REQUEST_POSTS_ERROR',
    payload: e,
  };
}

function receivePosts(data: {posts: Array<Post>}): Action {
  return {
    type: 'RECEIVE_POSTS',
    payload: data.posts,
    meta: {
      receivedAt: Date.now(),
    },
  };
}

function fetchPosts() {
  return (dispatch: Dispatch, getState: Function) => {
    if (getState().posts.isFetching) {
      return null;
    }
    dispatch(requestPosts());
    return PostsAPI.get()
      .then((response) => {
        if (!response.ok) {
          throw new Error(`Response Status ${response.status}`);
        }
        return response.json();
      })
      .then(json => dispatch(receivePosts(json)))
      .catch(e => dispatch(requestPostsError(e)));
  };
}

export {
  addPost,
  requestPosts,
  requestPostsError,
  receivePosts,
  fetchPosts,
};
