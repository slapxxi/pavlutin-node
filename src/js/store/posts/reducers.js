// @flow
import type { PostsState as State, Action } from '../flow-types';

const initialState: State = {
  isFetching: false,
  items: [],
  lastUpdated: 0,
};

function posts(posts: State = initialState, action: Action): State {
  switch (action.type) {
    case 'ADD_POST':
      return Object.assign({}, posts, {
        items: [...posts.items, action.payload],
        lastUpdated: Date.now(),
      });
    case 'REQUEST_POSTS':
      return Object.assign({}, posts, {
        isFetching: true,
      });
    case 'REQUEST_POSTS_ERROR':
      return Object.assign({}, posts, {
        isFetching: false,
      });
    case 'RECEIVE_POSTS':
      return Object.assign({}, posts, {
        items: action.payload.reverse(),
        lastUpdated: action.meta.receivedAt,
        isFetching: false,
      });
    default:
      return posts;
  }
}

export default posts;
