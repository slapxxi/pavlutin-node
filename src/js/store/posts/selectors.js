// @flow
import { find } from 'lodash';
import { createSelector } from 'reselect';
import type { State, Post } from '../flow-types';

function postsItems(state: State) {
  return state.posts.items;
}

function searchTerm(state: State) {
  return state.searchTerm;
}

function postsByTag(state: State, props: Object): Array<Post> {
  return filterByTag(state.posts.items, props.match.params.tag);
}

function postBySlug(state: State, props: Object): Post {
  return findPost(state.posts.items, props.match.params.slug);
}

function filterByTag(posts: Array<Post>, tag: string): Array<Post> {
  return !tag ? posts : posts.filter(p => p.tags.includes(tag));
}

function search(posts: Array<Post>, searchTerm: string): Array<Post> {
  if (searchTerm.trim() === '') {
    return posts;
  }
  const terms = searchTerm
    .trim()
    .split(' ')
    .filter(t => t.length >= 3)
    .map(term => new RegExp(term, 'i'));
  return posts.filter((post) => {
    const description = post.description || '';
    return !!find(terms, term => post.title.match(term) || description.match(term));
  });
}

function findPost(posts: Array<Post>, slug: string) {
  return find(posts, p => p.slug === slug);
}

const searchPosts = createSelector(
  [postsItems, searchTerm],
  search,
);

const searchPostsWithTag = createSelector(
  [postsByTag, searchTerm],
  search,
);

export { searchPosts, searchPostsWithTag, postBySlug };
