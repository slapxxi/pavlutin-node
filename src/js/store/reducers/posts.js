import types from '../actions/types';

function posts(
  posts = {
    isFetching: false,
    lastUpdated: 0,
    items: [],
  },
  action,
) {
  switch (action.type) {
    case types.ADD_POST:
      return Object.assign({}, posts, {
        items: [...posts.items, action.post],
        lastUpdated: Date.now(),
      });
    case types.REQUEST_POSTS:
      return Object.assign({}, posts, {
        isFetching: true,
      });
    case types.RECEIVE_POSTS:
      return Object.assign({}, posts, {
        items: action.posts.reverse(),
        lastUpdated: action.receivedAt,
        isFetching: false,
      });
    default:
      return posts;
  }
}

export default posts;
