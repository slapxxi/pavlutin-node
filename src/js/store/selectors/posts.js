import { find } from 'lodash';
import { createSelector } from 'reselect';

function postsItems(state) {
  return state.posts.items;
}

function searchTerm(state) {
  return state.searchTerm;
}

function postsByTag(state, props) {
  return filterByTag(state.posts.items, props.match.params.tag);
}

function postBySlug(state, props) {
  return findPost(state.posts.items, props.match.params.slug);
}

function filterByTag(posts, tag) {
  return !tag ? posts : posts.filter(p => p.tags.includes(tag));
}

function search(posts, searchTerm) {
  if (searchTerm.trim() === '') {
    return posts;
  }
  const terms = searchTerm
    .trim()
    .split(' ')
    .filter(t => t.length >= 3)
    .map(term => new RegExp(term, 'i'));
  return posts.filter((post) => {
    const description = post.description || '';
    return !!find(terms, term => post.title.match(term) || description.match(term));
  });
}

function findPost(posts, slug) {
  return find(posts, p => p.slug === slug);
}

const searchPosts = createSelector(
  [postsItems, searchTerm],
  search,
);

const searchPostsWithTag = createSelector(
  [postsByTag, searchTerm],
  search,
);

export { searchPosts, searchPostsWithTag, postBySlug };
