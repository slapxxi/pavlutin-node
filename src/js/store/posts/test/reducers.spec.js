import posts from '../reducers';
import {
  addPost,
  requestPosts,
  requestPostsError,
  receivePosts,
} from '../actions';

const post = { title: 'A first post' };

beforeAll(() => {
  Date.now = jest.fn(() => 100100100);
});

it('returns initial state', () => {
  const state = posts(undefined, {});
  expect(state).toMatchSnapshot();
});

it('handles adding post', () => {
  const state = posts(undefined, addPost(post));
  expect(state).toMatchSnapshot();
});

it('handles receiving posts', () => {
  const state = posts(
    undefined,
    receivePosts({ posts: [post] }),
  );
  expect(state).toMatchSnapshot();
});

it('handles requesting posts', () => {
  const state = posts({ items: [] }, requestPosts());
  expect(state).toMatchSnapshot();
});

it('handles request errors', () => {
  const state = posts(
    { items: [], isFetching: true },
    requestPostsError(),
  );
  expect(state).toMatchSnapshot();
});
