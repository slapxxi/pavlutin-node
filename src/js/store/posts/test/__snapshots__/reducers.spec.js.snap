// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles adding post 1`] = `
Object {
  "isFetching": false,
  "items": Array [
    Object {
      "title": "A first post",
    },
  ],
  "lastUpdated": 100100100,
}
`;

exports[`handles receiving posts 1`] = `
Object {
  "isFetching": false,
  "items": Array [
    Object {
      "title": "A first post",
    },
  ],
  "lastUpdated": 100100100,
}
`;

exports[`handles request errors 1`] = `
Object {
  "isFetching": false,
  "items": Array [],
}
`;

exports[`handles requesting posts 1`] = `
Object {
  "isFetching": true,
  "items": Array [],
}
`;

exports[`returns initial state 1`] = `
Object {
  "isFetching": false,
  "items": Array [],
  "lastUpdated": 0,
}
`;
